lib_xua/src/dfu/dfu.xc

```diff
@@ -147,54 +132,59 @@ static int DFU_Dnload(unsigned int request_len, unsigned int block_num, const un
     if (request_len == 0)
     {
         // Host signalling complete download
-        int i = 0;
-        unsigned int cmd_data[16];
         if (subPagesLeft)
         {
             unsigned int subPagePad[16] = {0};
-            for (i = 0; i < subPagesLeft; i++)
+            for (unsigned i = 0; i < subPagesLeft; i++)
             {
                 flash_cmd_write_page_data((subPagePad, unsigned char[64]));
             }
         }
-
-        cmd_data[0] = 2; // Terminate write
-        flash_cmd_write_page((cmd_data, unsigned char[]));
-
+        flash_cmd_end_write_image();
         DFU_state = STATE_DFU_MANIFEST_SYNC;
     }
     else
     {
-        unsigned int i = 0;
-        unsigned int flash_cmd = 0;
-        unsigned int flash_page_index = 0;
-        unsigned int cmd_data[16];
-
-        if (fromDfuIdle)
+        DFU_state = STATE_DFU_DOWNLOAD_SYNC; //from the spec. dfuDNLOAD-SYNC = Device has received a block and is waiting for the host to
+        // solicit the status via DFU_GETSTATUS. So if the host were to do a GetState right after this, it should see the device state as STATE_DFU_DOWNLOAD_SYNC.
+        // That is why, even when flash_cmd_start_write_image() returns not complete, we don't transition to STATE_DFU_DOWNLOAD_BUSY at this point but do it only
+        // from DFU_GetStatus()
+        if (!(block_num % 4)) // Every 4th block
         {
-            unsigned s = 0;
-
-            // Erase flash on first block
-            flash_cmd_erase_all();
-        }
-
-        // Program firmware, STATE_DFU_DOWNLOAD_BUSY not currently used
-        if (!(block_num % 4))
-        {
-            cmd_data[0] = !fromDfuIdle; // 0 for first page, 1 for other pages.
-            flash_cmd_write_page((cmd_data, unsigned char[64]));
+            flash_cmd_reset_subpage_index();
             subPagesLeft = 4;
+            if (fromDfuIdle) // Only relevant for block 0 which is when fromDfuIdle is also true
+            {
+                // Erase flash on block 0
+                flash_cmd_erase_all();
+
+                flash_cmd_start_write_image_in_progress = flash_cmd_start_write_image();
+#if DO_BLOCKING_ERASE
+                while(flash_cmd_start_write_image_in_progress)
+                {
+                    flash_cmd_start_write_image_in_progress = flash_cmd_start_write_image();
+                }
+#else
+                if(flash_cmd_start_write_image_in_progress) // flash_cmd_start_write_image() still in progress
+                {
+                    for (unsigned i = 0; i < 16; i++)
```

Comment:
I know this isn't your fault, but the 16 hardcoded everywhere is a bit nasty. sizeof(request data)/sizeof(requestdata[0]) would be more generic.

Reply:
Done. Cleaned up lots of other hardcoded values also!